### ISSUE TITLE
Race conditions in indexer cause constraint violations

### ISSUE THREAD

**Issue by goodsoft** [poanetwork/blockscout#1911]
I was investigating `status` constraint violations preventing internal transaction fetching for multiple blocks, and I found that quite complex race conditions occur due to asynchronous nature of indexer components.

Here's are three blocks fetched on 2019-05-07:
- 7713742n / `0x70ad961ff04b0e12551d5630822c465ccbfe31f13a91df5971782a6d42b926b1` / 127 txs – non-consensus version of block 7713742
- 7713742 / `0x2c83f6d0be5a2748a8f7e805ba3051b98add2c25674cbfb19cdb214450fe5bd3` / 95 txs
- 7713743 / `0x9eadba9fe3ad40a493666b0308e64a572dc93611c1db82ab4b36b57afdc54a8f`

Here's a timeline of their import:
- 12:46:12.412 – block 7713742n started importing
- 12:46:12.630924 – timestamp for block 7713742n, actual ordering undefined
- 12:46:13.629 – block 7713743 started importing
- 12:46:13.734798 – timestamp for block 7713743, actual ordering undefined
- 12:46:18.639 – block 7713742 started importing
- 12:48:19.249 – block 7713743 done importing
- 12:48:35.698 – block 7713742n done importing
- 12:49:04.715064 – multiple transactions added with 7713742n got _forked_ and had their `block_hash`/`block_number` erased
- ??:??:?? – these transactions got marked as `dropped/replaced` by replaced transactions worker due to their `block_hash` being `nil`
- 12:54:39.151 – block 7713742 failed importing due to database timeout at step `derive_address_current_token_balances`

- 13:20:14.690 – block 7713742 started importing
- 13:20:15.043172 – timestamp for block 7713742
- 13:22:23.265006 - internal transactions for block 7713742 imported
- 13:22:25.185 – block 7713742 done importing
- 13:22:30.027320 – internal transactions for block 7713742 re-imported
- 13:55:52.152542 – internal transactions for block 7713743 imported

Here's a supposed timeline of a particular transaction `0x04abeb35d618e9908641c36659ebbc49c17a4d1e1033722fd2ba8df89d204fbe`, which was present in 7713742n and ultimately in 7713743:
- imported with 7713743 (timestamp 12:46:13.734798)
- imported (overwritten!) with 7713742n (timestamp 12:46:12.630924)
- forked by failed import of 7713742 (timestamp 12:49:04.715064)
- marked as `dropped/replaced` by replaced transaction fetcher
- 13:55:52.152542 – internal transactions are fetched from 7713743, transaction's status can't be updated due to constraint `status` violation

Basically, the transaction was successfully fetched from a consensus block, later overwritten by other non-consensus block, and discarded by fork/replaced transaction handlers later.
As the consensus block wasn't re-fetched, the transaction was left in limbo.

**Comment by zachdaniel** [comment]
Perhaps a decent resolution to this would be to just reimport blocks when we detect potential dropped/replaced transactions? dropped/replaced transactions are signals that there was an uncle, right? That would mean that, if we ever detect them, we can just figure out which one was part of a consensus block (regardless of the timestamps of the transactions), and mark the non-consensus ones as dropped/replaced and reimport the consensus block?

**Comment by goodsoft** [comment]
Our `dropped/replaced` mark is actually for discarded _pending_ transactions, and it should actually be set only for transactions that were never collated.
Actually, the only problem with that is that we have missing `and is_nil(pending.block_hash)` in `Chain.find_and_update_replaced_transactions`, which might cause race condition problems.
Other than that, even if it is erroneously marked as `dropped/replaced` it would still be correctly refetched _if we trigger that refetch_.

Now the actual problem is triggering this refetch.
I think, we need to detect the moment we overwrite a collated transaction in `Explorer.Chain.Import.Runner.Transactions`, and trigger a refetch of the block it was originally in.
This way, even if a bad non-consensus block accidentally overwrites a good consensus one, we will just refetch the latter and justice will prevail.
